app/services/hashtag_autocomplete_service.rb

```diff
@@ -9,43 +9,55 @@ def self.search_conditions
   end
 
   attr_reader :guardian
-  cattr_reader :data_sources, :contexts
 
-  def self.register_data_source(type, klass)
-    @@data_sources[type] = klass
+  def self.data_sources
+    # Category and Tag data sources are in core and always should be
+    # included for searches and lookups.
+    [CategoryHashtagDataSource, TagHashtagDataSource] |
+      DiscoursePluginRegistry.hashtag_autocomplete_data_sources
   end
 
-  def self.clear_registered
-    @@data_sources = {}
-    @@contexts = {}
+  def self.contextual_type_priorities
+    # Category and Tag type priorities for the composer are default and
+    # always are included.
+    [
+      { type: "category", context: "topic-composer", priority: 100 },
+      { type: "tag", context: "topic-composer", priority: 50 },
+    ] | DiscoursePluginRegistry.hashtag_autocomplete_contextual_type_priorities
+  end
+
+  def self.data_source_types
+    data_sources.map(&:type)
+  end
 
-    register_data_source("category", CategoryHashtagDataSource)
-    register_data_source("tag", TagHashtagDataSource)
+  def self.data_source_icons
+    data_sources.map(&:icon)
+  end
 
-    register_type_in_context("category", "topic-composer", 100)
-    register_type_in_context("tag", "topic-composer", 50)
+  def self.data_source_from_type(type)
+    data_sources.find { |ds| ds.type == type }
   end
 
-  def self.register_type_in_context(type, context, priority)
-    @@contexts[context] = @@contexts[context] || {}
-    @@contexts[context][type] = priority
+  def self.find_priorities_for_context(context)
+    contextual_type_priorities.select { |ctp| ctp[:context] == context }
   end
 
-  def self.data_source_icons
-    @@data_sources.values.map(&:icon)
+  def self.unique_contexts
+    contextual_type_priorities.map { |ctp| ctp[:context] }.uniq
   end
 
   def self.ordered_types_for_context(context)
-    return [] if @@contexts[context].blank?
-    @@contexts[context].sort_by { |param, priority| priority }.reverse.map(&:first)
+    return [] if find_priorities_for_context(context).blank?
+    find_priorities_for_context(context)
+      .sort_by { |ctp| ctp[:priority] }
+      .reverse
+      .map { |ctp| ctp[:type] }
```

Comment:
Since `find_priorities_by_context` always returns an array, we can skip the guard clause.

```suggestion
    find_priorities_for_context(context)
      .sort_by { |ctp| ctp[:priority] }
      .reverse
      .map { |ctp| ctp[:type] }
```

Reply:
Yep will do :+1: